### DISCUSSION TITLE
[Question] failing cargo install in docker container

### DISCUSSION THREAD

**Discussion by kameshsampath** [rust-cross/cargo-zigbuild#92] [Q&A] [answered]
@messense  thank you so much for making this project. 

I am rust-lang newbie, trying to install `cargo install cargo-zigbuild` in `alpine:3.17` docker container. The cargo install fails technically getting killed, I also don't see any logs :( .  the same steps works on a `arm64` setup.

any thoughts here is appreciated.

**Comment by messense** [comment]
> trying to install `cargo install cargo-zigbuild` in `alpine:3.17` docker container.

> the same steps works on a `arm64` setup.

What architecture is the `alpine:3.17` docker container and what's your system cpu architecture?

`cargo-zigbuild` does cross compilation so you shouldn't install it in a QEMU emulated docker container.

**Reply by kameshsampath** [reply]
> > trying to install `cargo install cargo-zigbuild` in `alpine:3.17` docker container.
> 
> > the same steps works on a `arm64` setup.
> 
> What architecture is the `alpine:3.17` docker container and what's your system cpu architecture?
> 
it is `aarch64` 

> `cargo-zigbuild` does cross compilation so you shouldn't install it in a QEMU emulated docker container.

My need was to build a container with zig/cargo-zigbuild installed which could be used in my CI pipeline container e.g https://github.com/kameshsampath/rust-hello-world/blob/main/.drone.yml#L38

Here are my findings and fixes. Please let know your thoughts, 

1. I tried using multi stage builds and  tried copying the `zig` and `cargo-zigbuild` from the  `messene/cargo-zigbuild` image,  and  found that the `cargo-zigbuild` in that image was dynamically linked so tried adding `alpine-sdk` but it still then was looking for some dependencies, sadlly not being a C/C++ expert I resorted to `Step 2` below

2. I tried to do the typical `cargo install cargo-zigbuild` as here https://github.com/kameshsampath/rust-zig-builder/blob/main/Dockerfile#L7-L8 and then copied it my custom image in the following stage. This kind of helped me in building application successfully for `linux/amd64` and `linux/arm64`. 

It is very difficult to find the right set of dependencies thats required rust for runtime/compile time. Until now I found we need `gcc` , `musl*`, `libc` which together installed using `alpine-sdk` or build-essential(debian), sadly thats bloating the image size. Need to investigate further to nail down right list of dependencies.


I also found some memory issue with Docker for Deskop when doing multi arch builds, where I was trying to build the build the `linux/arm64` and `linux/amd64`. https://github.com/kameshsampath/rust-zig-builder. Thats when I see the `cargo install cargo-zigbuild` taking lots of time and fails.

Please let me know your thoughts.

**Comment by kameshsampath** [comment, marked as answer]
@messense 

Today I tried building it under a lima-vm with 4 CPU and 8GB of RAM and I get the following error for `cargo install cargo-zigbuild` 

```[ 2539.766683] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=/,mems_allowed=0,global_oom,task_memcg=/,task=cargo,pid=3478,uid=0
[ 2539.766705] Out of memory: Killed process 3478 (cargo) total-vm:8043844kB, anon-rss:7748312kB, file-rss:0kB, shmem-rss:2116kB, UID:0 pgtables:18508kB oom_score_adj:0
```

**Reply by kameshsampath** [reply]
not sure why install/compile should take so much memory . am I missing something

**Reply by messense** [reply]
Probably https://github.com/rust-lang/cargo/issues/10781?

**Reply by kameshsampath** [reply]
> Probably [rust-lang/cargo#10781](https://github.com/rust-lang/cargo/issues/10781)?

Thank you @messense  that really helped.
